Source/WarpX.cpp

```diff
@@ -1222,6 +1227,38 @@ WarpX::ReadParameters ()
             macroscopic_solver_algo = GetAlgorithmInteger(pp_algo,"macroscopic_sigma_method");
         }
 
+        if (evolve_scheme == EvolveScheme::ImplicitPicard ||
+            evolve_scheme == EvolveScheme::SemiImplicitPicard) {
+            utils::parser::queryWithParser(pp_algo, "max_picard_iterations", max_picard_iterations);
+            utils::parser::queryWithParser(pp_algo, "picard_iteration_tolerance", picard_iteration_tolerance);
+            utils::parser::queryWithParser(pp_algo, "require_picard_convergence", require_picard_convergence);
+
+            WARPX_ALWAYS_ASSERT_WITH_MESSAGE(
+                current_deposition_algo == CurrentDepositionAlgo::Esirkepov ||
+                current_deposition_algo == CurrentDepositionAlgo::Direct,
+                "Only Esirkepov or Direct current deposition supported with the implicit and semi-implicit schemes");
+
+            WARPX_ALWAYS_ASSERT_WITH_MESSAGE(
+                electromagnetic_solver_id == ElectromagneticSolverAlgo::Yee ||
+                electromagnetic_solver_id == ElectromagneticSolverAlgo::CKC,
+                "Only the Yee EM solver is supported with the implicit and semi-implicit schemes");
+
+            WARPX_ALWAYS_ASSERT_WITH_MESSAGE(
+                particle_pusher_algo == ParticlePusherAlgo::Boris,
```

Comment:
My impression is that the exact energy conservation would be preserved with any pusher for which the `vxB` force does not modify the norm of `u`, when updating `u` with `u^{n+1} - u^n = q*dt/m * (E + vxB)`. Using a dot product with `(u^{n+1} + u^n)`, we can see that this happens if `(u^{n+1} + u^n).(vxB) = 0`. In particular, this works if `v` is colinear with `u^{n+1} + u^n`, which is the case for the Boris and [Higuera-Carey]
(https://pubs.aip.org/aip/pop/article/24/5/052104/989750/Structure-preserving-second-order-integration-of) pusher (but not for the Vay pusher)
```suggestion
                particle_pusher_algo == ParticlePusherAlgo::Boris ||
                particle_pusher_algo == ParticlePusherAlgo::HigueraCarey
```

Reply:
Yes, the Higuera-Cary algorithm does work - energy is conserved.